File: solutions/graphs/bfs/cloneGraph.js
const cloneGraph = function (node) {
    if (!node) {
        return node;
    }

    const queue = new Queue();
    queue.enqueue(node);

    const originalToCloneMap = {};
    originalToCloneMap[node.val] = new Node(node.val, []);

    while (queue.size() > 0) {
        // Remove from queue
        const curNode = queue.dequeue();

        // Process node
        const curNodeNeighbors = curNode.neighbors;

        // Add neighbors
        for (const neighbor of curNodeNeighbors) {
            const hasNeighborClone = originalToCloneMap.hasOwnProperty(
                neighbor.val,
            );

            // This is an undirected graph, so cycles are definitely possible. This means that
            // we could traverse this graph forever, meaning we must keep track of which nodes
            // we visit. This `!hasNeighborClone` check is implicitly checking if the current
            // neighbor node has already been visited. This is because any time we visit a node,
            // we create an entry for it in `originalToCloneMap`. Thus, if `originalToCloneMap`
            // has no key for `neighbor.val` then we know that the current neighbor has NOT
            // been visited yet. So, this functions as our visited check.
            if (!hasNeighborClone) {
                queue.enqueue(neighbor);

                // Note, for each node, we must create an entry in our originalToCloneMap
                // BEFORE we process that node. This is so that other nodes are able to
                // add its cloned version as a neighbor. For example, suppose we have [1,2].
                // At node 1, we'll want to do something like node1Clone.neighbors.push(node2Clone).
                // Thus, we must create node2Clone when processing node1, we cannot just wait until
                // we process node2.
                originalToCloneMap[neighbor.val] = new Node(neighbor.val, []);
            }

            const curNodeClone = originalToCloneMap[curNode.val];
            const neighborClone = originalToCloneMap[neighbor.val];

            curNodeClone.neighbors.push(neighborClone);
        }
    }

    return originalToCloneMap[node.val];
};

File: solutions/python/graphs/bfs/clone_graph.py
class Solution:    
    def cloneGraph(self, start_node: 'Node') -> 'Node':
        if start_node is None:
            return start_node
        
        queue = deque()
        queue.append(start_node)
        
        original_to_clone_map = {}
        original_to_clone_map[start_node.val] = Node(start_node.val, [])
        
        while queue:
            # Remove node
            node = queue.popleft()
            
            # Process node
            # Note: there is no process node step for this problem.
            
            # Add neighbors            
            for neighbor in node.neighbors:
                if neighbor.val not in original_to_clone_map:
                    queue.append(neighbor)
                    original_to_clone_map[neighbor.val] = Node(neighbor.val, [])
                
                cur_node_clone = original_to_clone_map[node.val]
                neighbor_clone = original_to_clone_map[neighbor.val]
                
                cur_node_clone.neighbors.append(neighbor_clone)
        
        return original_to_clone_map[start_node.val]